### PR TITLE
Placeholder for CircleCI

### PR DESCRIPTION
**Pull request by andreineculau**
Concluding some minimum exploration of CircleCI as an alternative for the current TravisCI. Here's some random minuses

*Caches*: CircleCI caches are "immutable", and there seems to be no easy way to go around that i.e. to make job `n` restore the cache of job `n-1` of the same workflow&branch. Clearing caches is also not a feature at the moment, so the only thing to do is bump up a "version" marker in the cache name.

*UI*: maybe it's a matter of getting used to it, but clicking through the UI just for the purpose of browsing the workflow-to-job-and-back info is not optimized

*Logs*: no way to download raw logs from UI. I guess there's an API for it?! Also there is no code folding feature, so the logs could be sometimes hard to parse visually.

*Config*: rather duplicative and hard to follow. While support-firecloud's config is an outlier, overusing yaml references was a must, and just for the purpose of having a PR flow and a nightly flow, they will still be needed. No support for encrypted variables, but not a biggie.

*Docs*: I would say low-to-medium quality, a mix of info from CircleCI 1.0 and 2.0 and 2.1. The configuration I wrote for support-firecloud is half baked by googling the "CircleCI Discuss" platform.

*Features/Community*: it looks like there are core features that people see in other CI services, which are lingering to be implemented/ported to CircleCI, while at the same time some "features" (e.g. immutable caches) exist without having public support.

*Debug*: SSH debugging is available, but CircleCI triggers another run of the job to debug, and only afterwards it prints the SSH info. Also, it looks like the machine is available for 2 hours, allowing for multiple SSH sessions (a minus for security).

**EDIT** PR: one cannot have different workflows for PRs without shell hacks https://discuss.circleci.com/t/workflows-pull-request-filter/14396

---

My 2penny: good alternative, but might be bumpy.

PS: nice to see so few Travis specifics. I just needed to export some env vars, and everything worked. So porting to another CI is mainly equivalent to rewriting the configuration around what to run, when to run and cache.

### CODE REVIEW

**Comment by andreineculau** [comment]
FYI @IanSavchenko @alsmyr

**Comment by IanSavchenko** [comment]
> UI: maybe it's a matter of getting used to it, but clicking through the UI just for the purpose of browsing the workflow-to-job-and-back info is not optimized

Had the same impression. Some UI workflows are somewhat broken and non-linear.

> Logs: no way to download raw logs from UI. I guess there's an API for it?! Also there is no code folding feature, so the logs could be sometimes hard to parse visually.

Yes, but not stupid log size limitation.

> Config: rather duplicative and hard to follow. While support-firecloud's config is an outlier, overusing yaml references was a must, and just for the purpose of having a PR flow and a nightly flow, they will still be needed. No support for encrypted variables, but not a biggie.

Tried it on a smaller project and didn't see much difference with Travis. 

> Docs: I would say low-to-medium quality, a mix of info from CircleCI 1.0 and 2.0 and 2.1. The configuration I wrote for support-firecloud is half baked by googling the "CircleCI Discuss" platform.

Yes, I also found the mix to be confusing. But everything seemed to be googlable.

As I understand, we can also leverage our own docker images for the builds, which is not possible on Travis.

**Comment by andreineculau** [comment]
> Yes, but not stupid log size limitation.

AFAICS, it's not without problems in this area

* https://discuss.circleci.com/t/view-or-download-full-output-logs/247
* https://discuss.circleci.com/t/uncapped-logging-output/3367

**Comment by andreineculau** [comment]
new negative discovery: circleci only supports notifying one slack room

https://discuss.circleci.com/t/can-i-enable-slack-notifications-to-multiple-rooms/453

**Comment by andreineculau** [comment]
other alternatives:

* https://buildkite.com/features
* https://drone.io
* https://buddy.works

**Comment by andreineculau** [comment]
On circleci: I subscribed to https://status.circleci.com notifications (exactly a month ago), and they are many. One could argue that maybe they are better, more meticulous, at monitoring than others, or that they monitor also side-things (like npm for instance), I don't know what it's the cause, but the feeling that I have is of "regular hiccups".

**Comment by andreineculau** [comment]
random find https://cirrus-ci.org supports linux, macos, win32, freebsd, docker containers...

**Comment by alsmyr** [comment]
Is it possible to see how much time to utilize Travis? In order to assess the cost on a pay per %time% model like Cirrus

**Comment by andreineculau** [comment]
sent on private channel
